gh-127174: add some advice for asyncio.get_event_loop replacements (#127640)
python/cpython: Doc/whatsnew/3.14.rst

* Removed implicit creation of event loop by :func:`asyncio.get_event_loop`.
  It now raises a :exc:`RuntimeError` if there is no current event loop.
  (Contributed by Kumar Aditya in :gh:`126353`.)
  There's a few patterns that use :func:`asyncio.get_event_loop`, most
  of them can be replaced with :func:`asyncio.run`.

>   If you're running an async function, simply use :func:`asyncio.run`,

Reviewer: ```suggestion
  If you're running an async function, simply use :func:`asyncio.run`.
  
```

* Removed implicit creation of event loop by :func:`asyncio.get_event_loop`.
  It now raises a :exc:`RuntimeError` if there is no current event loop.
  (Contributed by Kumar Aditya in :gh:`126353`.)
  There's a few patterns that use :func:`asyncio.get_event_loop`, most
  of them can be replaced with :func:`asyncio.run`.

  If you're running an async function, simply use :func:`asyncio.run`.
  
  before::

    async def main():
        ...

    loop = asyncio.get_event_loop()
    try:
       loop.run_until_complete(amain())
    finally:
       loop.close()

  after::

    async def main():
        ...

    asyncio.run(main())

  If you need to start something, e.g. a server listening on a socket
  and then run forever, use :func:`asyncio.run` and an

>   :class:`asyncio.Event`, before::

Reviewer: ```suggestion
  :class:`asyncio.Event`.
  
  before::
```

* Removed implicit creation of event loop by :func:`asyncio.get_event_loop`.
  It now raises a :exc:`RuntimeError` if there is no current event loop.
  (Contributed by Kumar Aditya in :gh:`126353`.)
  There's a few patterns that use :func:`asyncio.get_event_loop`, most
  of them can be replaced with :func:`asyncio.run`.

  If you're running an async function, simply use :func:`asyncio.run`.
  

>   before::

Reviewer 2: ```suggestion
  Before::
```

    async def main():
        ...

    loop = asyncio.get_event_loop()
    try:

>        loop.run_until_complete(amain())

Reviewer 2: Can you use 4-spaces indents in the Python code please?

    finally:
       loop.close()

>   after::

Reviewer 2: ```suggestion
  After::
```

>     async def amain():
>         ...
> 
>     loop = asyncio.get_event_loop()
>     try:
>        loop.run_until_complete(amain())
>     finally:
>        loop.close()
> 
>   after::
> 
>     async def amain():
>         ...
> 
>     asyncio.run(amain())

Reviewer: ```suggestion
    async def main():
        ...

    loop = asyncio.get_event_loop()
    try:
       loop.run_until_complete(amain())
    finally:
       loop.close()

  after::

    async def main():
        ...

    asyncio.run(main())
```

>   If you need to start something, eg, a server listening on a socket

Reviewer: ```suggestion
  If you need to start something, e.g. a server listening on a socket
```

>   and then run forever, use :func:`asyncio.run` and an

Reviewer: I don't think this server example is useful to add, I'm fine with the Runner one though

Author: This is a really common pattern that I see, and people just don't know they can wait for an Event forever to get run_forever behavior

Reviewer: I feel that it really belongs to an asyncio recipes section rather than whatnew page, this pattern orthogonal to the change listed here. What do you think?

Author: I really think it's useful here, perhaps it can be added to recipes later?

Eg for users of 3.15+

  and then run forever, use :func:`asyncio.run` and an
  :class:`asyncio.Event`.
  

>   before::

Reviewer 2: ```suggestion
  Before::
```

    def start_server(loop):
        ...

    loop = asyncio.get_event_loop()
    try:
        start_server(loop)
        loop.run_forever()
    finally:
        loop.close()

>   after::

Reviewer 2: ```suggestion
  After::
```

>     async def amain():

Reviewer 3: Why switch to `amain()` instead of `main()` like before? I think all examples should use a consistent style.

```suggestion
    async def main():
```

Author: Yeah I prefer the amain, it's compatible with idiomatic usage of console scripts:

```python
import asyncio
import sys

async def amain():
    ...

def main():
    return asyncio.run(amain())

if __name__ == "__main__":
    sys.exit(main())
```

        start_server(asyncio.get_running_loop())
        await asyncio.Event().wait()

>     asyncio.run(amain())

Reviewer 3: ```suggestion
    asyncio.run(main())
```

  If you need to run something in an event loop, then run some blocking

>   code around it, use :class:`asyncio.Runner`, before::

Reviewer 3: ```suggestion
  code around it, use :class:`asyncio.Runner`.
```

Author: This needs the :: for the code block to work

> 

Reviewer 3: ```suggestion

before::

```

    async def operation_one():
        ...

    def blocking_code():
        ...

    async def operation_two():
        ...

    loop = asyncio.get_event_loop()
    try:
       loop.run_until_complete(operation_one())
       blocking_code()
       loop.run_until_complete(operation_two())
    finally:
       loop.close()

  after::

    async def operation_one():
        ...

    def blocking_code():
        ...

    async def operation_two():
        ...

    with asyncio.Runner() as runner:
       runner.run(operation_one())
       blocking_code()
       runner.run(operation_two())

>   If you need an 'ambient' event loop, shared between multiple otherwise

Reviewer: Remove this part

>   If you need an 'ambient' event loop, shared between multiple otherwise

Author: ```suggestion
```

>   unrelated modules, this is the sort of use-case that's no longer supported

Author: ```suggestion
```

>   in asyncio.

Author: ```suggestion
```
